src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,194 @@
+                {
+                    return;
+                }
+
+                if (!(descendant is TIdentifierNameSyntax name))
+                {
+                    continue;
+                }
+
+                var symbol = model.GetSymbolInfo(descendant).Symbol as IFieldSymbol;
+                if (symbol == null || !unassignedSymbols.Contains(symbol))
+                {
+                    continue;
+                }
+
+                if (!IsMemberOfThisInstance(descendant))
+                {
+                    unassignedSymbols.Remove(symbol);
+                }
+
+                if (IsDescendentOf<TConstructorDeclarationSyntax>(descendant, out var ctorNode))
+                {
+                    var isInAnonymousOrLocalFunction = false;
+                    for (var current = descendant.Parent; current != ctorNode; current = current.Parent)
+                    {
+                        if (syntaxFactsService.IsAnonymousOrLocalFunction(current))
+                        {
+                            isInAnonymousOrLocalFunction = true;
+                            break;
+                        }
+                    }
+
+                    if (isInAnonymousOrLocalFunction)
+                    {
+                        unassignedSymbols.Remove(symbol);
+                    }
+                    else
+                    {
+                        var ctorSymbol = model.GetDeclaredSymbol(ctorNode);
+                        if (!ctorSymbol.ContainingType.Equals(symbol.ContainingType))
+                        {
+                            unassignedSymbols.Remove(symbol);
+                        }
+
+                        if (!ctorSymbol.IsStatic && symbol.IsStatic)
+                        {
+                            unassignedSymbols.Remove(symbol);
+                        }
+                    }
+
+                    // assignments in the ctor don't matter other than the static modifiers and lambdas point checked above
+                    continue;
+                }
+                
+                if (IsWrittenTo(name, model, cancellationToken))
+                {
+                    unassignedSymbols.Remove(symbol);
+                }
+            }
+        }
+        
+        private bool IsDescendentOf<T>(SyntaxNode node, out T ctor) where T : SyntaxNode
+        {
+            ctor = node.FirstAncestorOrSelf<T>();
+            return ctor != null;
+        }
+
+        private bool IsMutableValueType(ITypeSymbol type)
+        {
+            if (type.TypeKind != TypeKind.Struct)
+            {
+                return false;
+            }
+
+            foreach (var member in type.GetMembers())
+            {
+                if (member is IFieldSymbol fieldSymbol &&
+                    !(fieldSymbol.IsConst || fieldSymbol.IsReadOnly || fieldSymbol.IsStatic))
+                {
+                    return true;
+                }
+            }
```

Comment:
there's a major problem/concern i have here.  This works if we can accurately get the members of hte struct.  But what does this mean for reference assemblies (where fields have been stripped out)?

Reply:
➡️ All fields are kept for `struct` types. https://github.com/dotnet/roslyn/blob/master/docs/features/refout.md#definition-of-ref-assemblies

Reply:
I made a comment in #19994 documenting our reliance on this behavior.